primefaces/src/main/java/org/primefaces/component/datatable/DataTableRenderer.java

```diff
@@ -980,11 +980,16 @@ public void encodeTbody(FacesContext context, DataTable table, boolean dataOnly,
         int rows = table.getRows();
         int first = table.isClientCacheRequest(context) ? Integer.parseInt(params.get(clientId + "_first")) + rows : table.getFirst();
         int rowCount = table.getRowCount();
-        int rowCountToRender = rows == 0 ? (table.isLiveScroll() ? (table.getScrollRows() + table.getScrollOffset()) : rowCount) : rows;
 
+        int rowCountToRender;
         if (table.isVirtualScroll()) {
-            int virtualScrollRowCount = (table.getScrollRows() * 2);
-            rowCountToRender = Math.min(virtualScrollRowCount, rowCount);
+            rowCountToRender = Math.min(table.getScrollRows() * 2, rowCount);
+        }
+        else if (table.isLiveScroll()) {
+            rowCountToRender = table.getScrollRows() + table.getScrollOffset();
+        }
+        else {
+            rowCountToRender = rows == 0 ? rowCount : rows;
         }
```

Comment:
I think this logic needs to be 

```java
int rowCountToRender;

if (rows == 0) {
    if (table.isLiveScroll()) {
        rowCountToRender = Math.min(table.getScrollRows() * 2, rowCount);
    }
    else if (table.isLiveScroll()) {
        rowCountToRender = table.getScrollRows() + table.getScrollOffset();
    }
    else {
        rowCountToRender = rowCount;
    }
} else {
    rowCountToRender = rows;
}
```

Reply:
the virtualScoll if was outside before
wait 1 min, maybe this is more clear

Reply:
Yeah it looks like they were specifically looking for 0 rows before attempting the other checks?

Reply:
not sure TBH
nows its 100% the old logic again